diff -wBbur ../org/decoders/timidity/timidity.c ./decoders/timidity/timidity.c
--- ../org/decoders/timidity/timidity.c	2008-04-17 10:56:20 -0700
+++ ./decoders/timidity/timidity.c	2014-07-26 15:57:02 -0700
@@ -27,6 +27,9 @@
 #include <stdlib.h>
 #include <string.h>
 
+#if defined (__MINGW32__)
+#include <windows.h>
+#endif
 #include "SDL_sound.h"
 
 #define __SDL_SOUND_INTERNAL__
@@ -424,6 +427,14 @@
    *            that needs to be added to the search path.
    */
 #ifdef WIN32
+#if defined (__MINGW32__)
+    char buf[255], *p;
+    if (SearchPath(NULL,"msys-2.0.dll",NULL,255,buf,NULL)) {
+	p = strstr(buf,"bin\\msys");
+	strcpy(p,"local/share/timidity");
+	add_to_pathlist(buf);
+    }
+#endif
   add_to_pathlist("\\TIMIDITY");
 #else
   add_to_pathlist("/usr/local/lib/timidity");
