diff -ru ../dosbox-code-0/include/bios_disk.h include/bios_disk.h
--- ../dosbox-code-0/include/bios_disk.h	2022-05-04 07:31:48.207160300 -0700
+++ include/bios_disk.h	2022-05-02 13:33:03.126182900 -0700
@@ -66,10 +66,15 @@
 	Bit32u sector_size;
 	Bit32u heads,cylinders,sectors;
 private:
-	Bit32u current_fpos;
+	Bit64u current_fpos;
 	enum { NONE,READ,WRITE } last_action;
 };
 
+#ifndef __APPLE__
+#define fseek fseeko64
+#define ftell ftello64
+#endif
+
 void updateDPT(void);
 void incrementFDD(void);
 
diff -ru ../dosbox-code-0/src/ints/bios_disk.cpp src/ints/bios_disk.cpp
--- ../dosbox-code-0/src/ints/bios_disk.cpp	2022-05-04 07:31:45.410194900 -0700
+++ src/ints/bios_disk.cpp	2022-05-02 13:33:03.134160100 -0700
@@ -159,9 +159,9 @@
 }
 
 Bit8u imageDisk::Read_AbsoluteSector(Bit32u sectnum, void * data) {
-	Bit32u bytenum;
+	Bit64u bytenum;
 
-	bytenum = sectnum * sector_size;
+	bytenum = (Bit64u)sectnum * sector_size;
 
 	if (last_action==WRITE || bytenum!=current_fpos) fseek(diskimg,bytenum,SEEK_SET);
 	size_t ret=fread(data, 1, sector_size, diskimg);
@@ -181,9 +181,9 @@
 
 
 Bit8u imageDisk::Write_AbsoluteSector(Bit32u sectnum, void *data) {
-	Bit32u bytenum;
+	Bit64u bytenum;
 
-	bytenum = sectnum * sector_size;
+	bytenum = (Bit64u)sectnum * sector_size;
 
 	//LOG_MSG("Writing sectors to %ld at bytenum %d", sectnum, bytenum);
 
